rust/src/domains/polars/frame/mod.rs

```diff
@@ -0,0 +1,429 @@
+        }
+        write!(
+            f,
+            "LazyFrameDomain({}{})",
+            self.series_domains
+                .iter()
+                .map(|s| format!("{}: {}", s.field.name, s.field.dtype))
+                .collect::<Vec<_>>()
+                .join(", "),
+            margins_debug
+        )
+    }
+}
+
+impl<F: Frame> Domain for FrameDomain<F> {
+    type Carrier = F;
+    fn member(&self, val: &Self::Carrier) -> Fallible<bool> {
+        let val_df = val.clone().dataframe()?;
+
+        // if df has different number of columns as domain
+        if val_df.schema().len() != self.series_domains.len() {
+            return Ok(false);
+        }
+
+        // check if each column is a member of the series domain
+        for (s, dom) in val_df.get_columns().iter().zip(self.series_domains.iter()) {
+            if !dom.member(s)? {
+                return Ok(false);
+            }
+        }
+
+        // check that margins are satisfied
+        for (by, margin) in self.margins.iter() {
+            if !margin.member(by, val.clone().lazyframe())? {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+/// A restriction on the unique values in the margin, as well as possibly their counts,
+/// over a set of columns in a LazyFrame.
+#[derive(Clone, PartialEq)]
+pub struct Margin {
+    /// The greatest number of records that can be present in any one partition.
+    pub max_partition_size: Option<u32>,
+    /// The greatest number of partitions that can be present.
+    pub max_num_partitions: Option<u32>,
+
+    /// The greatest number of contributions that can be made by one unit to any one partition.
+    ///
+    /// This affects how margins interact with the metric.
+    /// The distance between data sets differing by more than this quantity is considered infinite.
+    pub max_partition_contributions: Option<u32>,
+    /// The greatest number of partitions that can be contributed to.
+    ///
+    /// This affects how margins interact with the metric.
+    /// The distance between data sets differing by more than this quantity is considered infinite.
+    pub max_influenced_partitions: Option<u32>,
+
+    /// Denote whether the unique values and/or in the margin are public.
+    pub public_info: Option<MarginPub>,
+}
+
+#[derive(Clone, PartialEq)]
+/// Denote how margins interact with the metric.
+pub enum MarginPub {
+    /// The distance between data sets with different margin keys are is infinite.
+    Keys,
+    /// The distance between data sets with different margin keys or partition lengths is infinite.
+    Lengths,
+}
+
+impl Margin {
+    pub fn new() -> Self {
+        Margin {
+            max_partition_size: None,
+            max_num_partitions: None,
+            max_partition_contributions: None,
+            max_influenced_partitions: None,
+            public_info: None,
+        }
+    }
+
+    pub fn max_partition_size(mut self, value: u32) -> Self {
```

Comment:
`with_xxx` is a really common way of writing Rust APIs. 
I think the compiler figures out how to inline it, and since rust doesn't have named arguments, it gives you a way to instantiate structs in a readable way without making an ocean of `None`'s.